src/command/subgraph/publish.rs

```diff
@@ -74,4 +126,84 @@ impl Publish {
             publish_response,
         })
     }
+
+    // output stream is optionally injected (defaulting to stdout) so we can test this function
+    pub fn handle_invalid_routing_url(
+        maybe_prompt_response: Option<bool>,
+        maybe_output: Option<&mut dyn std::io::Write>,
+    ) -> Option<RoverOutput> {
+        let default_output = &mut std::io::stdout();
+        let output = maybe_output.unwrap_or(default_output);
+        if let Some(prompt_response) = maybe_prompt_response {
+            if prompt_response {
+                None
+            } else {
+                writeln!(output, "Publish cancelled by user")
+                    .expect("Could not write to provided output stream");
+                Some(RoverOutput::EmptySuccess)
+            }
+        } else {
+            // if we're not in a tty, we didn't prompt. let's print a warning but
+            // publish anyway.
+            writeln!(
+                output,
+                "{} In a future major version of Rover, the `--allow-invalid-routing-url` flag will be required to publish a subgraph with an invalid routing URL in CI.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            writeln!(
+                output,
+                "{} Found an invalid URL, but we can't prompt in a non-interactive environment. Publishing anyway.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            None
+        }
+    }
+
+    pub fn prompt_for_publish() -> Option<bool> {
+        if !atty::is(atty::Stream::Stdout) {
+            // q for Avery: not actually sure why I can't implicitly return
+            // `None` here
+            return None;
+        }
+
+        match prompt_confirm_default_no(
+            "Found an invalid URL, would you still like to publish? [y/N]: ",
```

Comment:
i'm not actually getting this to print the warning or prompt me with invalid `--routing-url`s, i haven't found where the bug is but this is the command i used to try to get the warning to show up

```console
$ cargo rover subgraph publish averys-supercloud@new-variant-on-existing-graph --name products --routing-url definitely-invalid --schema ./examples/supergraph-demo/products/products.graphql
A new subgraph called 'products' was created in 'averys-supercloud@new-variant-on-existing-graph'
The supergraph schema for 'averys-supercloud@new-variant-on-existing-graph' was updated, composed from the updated 'products' subgraph
Monitor your schema delivery progression on studio: https://studio.apollographql.com/graph/averys-supercloud/launches/35a69868-ff2e-41c6-ada9-a942ac1e8c02?variant=new-variant-on-existing-graph
```

Reply:
Hmm, I'm not reproducing using the URL you did. Seems to be working (though I spotted that I'm duping the y/N part of the prompt). We can investigate tomorrow.